linux/devdoc/threadpool_linux_requirements.md

```diff
@@ -0,0 +1,258 @@
+
+`threadpool_destroy` shall destroy the semphore by calling `sem_destroy`.
+
+`threadpool_destroy` shall destroy the SRW lock by calling `srw_lock_destroy`.
+
+### threadpool_open_async
+
+```C
+MOCKABLE_FUNCTION(, int, threadpool_open_async, THREADPOOL_HANDLE, threadpool, ON_THREADPOOL_OPEN_COMPLETE, on_open_complete, void*, on_open_complete_context);
+```
+
+`threadpool_open_async` opens the threadpool asynchronously.
+
+If `threadpool` is `NULL`, `threadpool_open_async` shall fail and return a non-zero value. 
+
+`threadpool_open_async` shall set the state to `THREADPOOL_STATE_OPENING`. 
+
+If `threadpool` has already been opened, `threadpool_open_async` shall fail and return a non-zero value.
+
+`threadpool_open_async` shall create the threads for `threadpool` using `ThreadAPI_Create`.
+
+If one of the thread creation fails, `threadpool_open_async` shall fail and return a non-zero value, terminate all threads already created, indicate an error to the user by calling the `on_open_complete` callback with `THREADPOOL_OPEN_ERROR` and set `threadpool` state to `THREADPOOL_STATE_NOT_OPEN`.
+
+Otherwise, `threadpool_open_async` shall set the state to `THREADPOOL_STATE_OPEN`, indicate open success to the user by calling the `on_open_complete` callback with `THREADPOOL_OPEN_OK` and return zero.
+
+### threadpool_close
+
+```C
+MOCKABLE_FUNCTION(, void, threadpool_close, THREADPOOL_HANDLE, threadpool);
+```
+
+`threadpool_close` closes the threadpool.
+
+If `threadpool` is `NULL`, `threadpool_close` shall fail and return.
+
+Otherwise, `threadpool_close` shall terminate all threads in the `threadpool` and set the state to `THREADPOOL_STATE_NOT_OPEN`.
+
+### threadpool_schedule_work
+
+```C
+MOCKABLE_FUNCTION(, int, threadpool_schedule_work, THREADPOOL_HANDLE, threadpool, THREADPOOL_WORK_FUNCTION, work_function, void*, work_function_context);
+```
+
+`threadpool_schedule_work` schedule the threadpool to work.
+
+If `threadpool` is `NULL`, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+If `work_function` is `NULL`, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+`threadpool_schedule_work` shall get the threadpool state by calling `interlocked_add`.
+
+`threadpool_schedule_work` shall validate that it's state is THREADPOOL_STATE_OPEN and if not shall fail and return a non-zero value.
+
+`threadpool_schedule_work` shall increment the count of pending call that are in progress to be executed.
+
+`threadpool_schedule_work` shall acquire the shared SRW lock by calling `srw_lock_acquire_shared`.
+
+`threadpool_schedule_work` shall increment the `insert_pos` by one.
+
+`threadpool_schedule_work` shall set the current task state to `TASK_INITIALIZING`.
+
+If task has been already initialized, `threadpool_schedule_work` shall release the shared SRW lock by calling `srw_lock_release_shared` and increase `task_array` capacity by calling `reallocate_threadpool_array`.
```

Comment:
>by calling `reallocate_threadpool_array`. [](http://example.com/codeflow?start=171&length=41)

The fact that the code is extracted in a function inside the module has no value for the spec. We're not meaning to reference the portion that reallocates multiple places in the spec, so we should not even mention it separately.

The actions to reallocate should just be described here.

Reply:
moved specs for reallocating the task array to here